vault/ttl.go

```diff
@@ -0,0 +1,90 @@
+package vault
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+func calculateTTL(sysView logical.SystemView, increment, backendTTL, period, backendMaxTTL, explicitMaxTTL time.Duration, startTime time.Time) (ttl time.Duration, warnings []string, errors error) {
+	// Truncate all times to the second since that is the lowest precision for
+	// TTLs
+	now := time.Now().Truncate(time.Second)
+	if startTime.IsZero() {
+		startTime = now
+	} else {
+		startTime = startTime.Truncate(time.Second)
+	}
+
+	// Start off with the sys default value, and update according to period/TTL
+	// from resp.Auth
+	ttl = sysView.DefaultLeaseTTL()
+
+	// Use the mount's configured max unless the backend specifies
+	// something more restrictive (perhaps from a role configuration
+	// parameter)
+	maxTTL := sysView.MaxLeaseTTL()
+	if backendMaxTTL > 0 && backendMaxTTL < maxTTL {
+		maxTTL = backendMaxTTL
+	}
+	if explicitMaxTTL > 0 && explicitMaxTTL < maxTTL {
+		maxTTL = explicitMaxTTL
+	}
+
+	// Should never happen, but guard anyways
+	if maxTTL < 0 {
+		return 0, nil, fmt.Errorf("max TTL is negative")
+	}
+
+	// Determine the correct increment to use
+	if increment <= 0 {
+		if backendTTL > 0 {
+			increment = backendTTL
+		} else {
+			increment = sysView.DefaultLeaseTTL()
+		}
+	}
+
+	var maxValidTime time.Time
+	switch {
+	case period > 0:
+		// Cap the period value to the sys max_ttl value
+		if period > maxTTL {
+			warnings = append(warnings,
+				fmt.Sprintf("Period of %q exceeded the effective max_ttl of %q; Period value is capped accordingly", period, maxTTL))
+			period = maxTTL
+		}
+		ttl = period
+
+		if explicitMaxTTL > 0 {
+			maxValidTime = startTime.Add(explicitMaxTTL)
+		}
+	case increment > 0:
+		// We cannot go past this time
+		maxValidTime = startTime.Add(maxTTL)
+		ttl = increment
+	}
+
+	if !maxValidTime.IsZero() {
+		// Determine the max valid TTL
+		maxValidTTL := maxValidTime.Sub(now)
```

Comment:
I think it should be `maxValidTIme.Sub(startTime)`

Reply:
maxValidTTL is the largest TTL that is valid from this point, not the total TTL valid since it was issued. Hence below, if the calculated TTL (which is from this point) is larger than the max valid TTL from this point, it has to be capped.

Reply:
I think it should be `now`.

Reply:
Ah, yeah this case we are comparing max time against current time.